### PR TITLE
feat(serde_v8): DetachedBuffer

### PR DESCRIPTION
**Pull request by AaronO**
This would be useful in WebGPU, where the spec's semantics (IIRC) expect buffers to be detached.

It might also be useful in other situations where we want to transfer ownership across the op-layer rather than share ownership

CC @lucacasonato @crowlKats

### CODE REVIEW

File: serde_v8/src/magic/zero_copy_buf.rs [hunk 1]
```diff
@@ -19,118 +19,87 @@ use super::transl8::FromV8;
...
-    scope: &mut v8::HandleScope,
-    view: v8::Local<v8::ArrayBufferView>,
-  ) -> Result<Self, v8::DataError> {
+  fn as_slice(&self) -> &[u8] {
+    unsafe { &*(&self.store[self.range.clone()] as *const _ as *const [u8]) }
+  }
+
+  #[allow(clippy::cast_ref_to_mut)]
+  fn as_slice_mut(&mut self) -> &mut [u8] {
+    unsafe { &mut *(&self.store[self.range.clone()] as *const _ as *mut [u8]) }
```
Comment (lucacasonato) [inline serde_v8/src/magic/zero_copy_buf.rs:49]: `// SAFETY`

File: serde_v8/src/magic/zero_copy_buf.rs [hunk 2]
```diff
@@ -19,118 +19,87 @@ use super::transl8::FromV8;
...
+    let (offset, len) = (view.byte_offset(), view.byte_length());
     let buffer = view.buffer(scope).ok_or(v8::DataError::NoData {
       expected: "view to have a buffer",
     })?;
-    Self::from_buffer(buffer, view.byte_offset(), view.byte_length())
+    let buffer = unsafe { std::mem::transmute(buffer) }; // lifetime hack
+    return Ok((buffer, offset..offset + len));
   }
+  let b: v8::Local<v8::ArrayBuffer> = value.try_into()?;
+  let b = unsafe { std::mem::transmute(b) }; // lifetime hack
```
Comment (lucacasonato) [inline serde_v8/src/magic/zero_copy_buf.rs:67]: `// SAFETY`

File: serde_v8/src/magic/zero_copy_buf.rs [hunk 3]
```diff
@@ -19,118 +19,87 @@ use super::transl8::FromV8;
...
+  value: v8::Local<'a, v8::Value>,
+) -> Result<(v8::Local<'a, v8::ArrayBuffer>, Range<usize>), v8::DataError> {
+  if value.is_array_buffer_view() {
+    let view: v8::Local<v8::ArrayBufferView> = value.try_into()?;
+    let (offset, len) = (view.byte_offset(), view.byte_length());
     let buffer = view.buffer(scope).ok_or(v8::DataError::NoData {
       expected: "view to have a buffer",
     })?;
-    Self::from_buffer(buffer, view.byte_offset(), view.byte_length())
+    let buffer = unsafe { std::mem::transmute(buffer) }; // lifetime hack
```
Comment (lucacasonato) [inline serde_v8/src/magic/zero_copy_buf.rs:63]: `// SAFETY`

**Comment by aapoalas** [comment]
Aside from the missing safety comments LGTM.

File: serde_v8/src/magic/zero_copy_buf.rs (hunk 3 above)
Comment (AaronO) [reply serde_v8/src/magic/zero_copy_buf.rs]: Removed unsafe in [50b41b5](https://github.com/denoland/deno/pull/14102/commits/50b41b5ba5f637f0dc21233c79c044da2477f8c9)

File: serde_v8/src/magic/zero_copy_buf.rs (hunk 2 above)
Comment (AaronO) [reply serde_v8/src/magic/zero_copy_buf.rs]: Removed unsafe in [50b41b5](https://github.com/denoland/deno/pull/14102/commits/50b41b5ba5f637f0dc21233c79c044da2477f8c9)

File: serde_v8/src/magic/zero_copy_buf.rs (hunk 1 above)
Comment (AaronO) [reply serde_v8/src/magic/zero_copy_buf.rs]: Scoped to https://github.com/denoland/deno/pull/14095, which should land before this PR

File: serde_v8/magic/detached_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+impl FromV8 for DetachedBuffer {
+  fn from_v8(
+    scope: &mut v8::HandleScope,
+    value: v8::Local<v8::Value>,
+  ) -> Result<Self, crate::Error> {
+    let (b, range) =
+      to_ranged_buffer(scope, value).or(Err(crate::Error::ExpectedBuffer))?;
+    let store = b.get_backing_store();
+    b.detach(); // Detach
```
Comment (andreubotella) [inline serde_v8/magic/detached_buffer.rs:63]: The backing buffers for Wasm memories are non-detachable (`!b.is_detachable()`). I'm not sure if V8 panics if you try to detach them, but that should probably return an error in serde_v8.

Comment (AaronO) [reply serde_v8/magic/detached_buffer.rs:63]: Yep, I was meaning to add an IsDetachable guard 👍

**Review by littledivy — APPROVED** [review]
LGTM
